### DISCUSSION TITLE
Report for quality of GPS data

### DISCUSSION THREAD

**Discussion by ph818** [OpenTracksApp/OpenTracks#949] [Ideas]
I'm finding that OpenTracks is the most transparent and currently maintained Android program for getting for GPX recordings for trail mapping, etc.

I wonder if open tracks could show information about the quality of the track - whether it detects energy saver mode interfering, when it had to discard location data because of low accuracy, etc, and save that as a log. It would help with troubleshooting poor-quality tracks. Is there already a way to get this information?

**Comment by dennisguse** [comment]
We store the horizontal accuracy as provided by Android.
If we discard a GPS location, this information is not stored though. We still lack the option to export this data, so it can be used in post processing with other tools.

PS: Battery saver is tough as then we don't get any GPS data at all.

**Comment by brianjmurrell** [comment]
Ahhh.  I was just going to start a discussion on exporting (some how) in a GPX file whether GPS was enabled for the track or not.  But it seems that this horizontal accuracy would be the solution.  This would also be usable for specifying the accuracy when uploading the data to Google Fit via the [com.google.location.sample](https://developers.google.com/fit/datatypes/location#location_sample)'s `accuracy` data item.

Here's an interesting [StackOverflow Q&A](https://stackoverflow.com/questions/5185823/specifying-accuracy-in-gpx-or-kml) about expressing accuracy in GPX files with a wikipedia link for _Dilution of precision_.  Other searches seem to indicate people using the `<hdop>` tag for this.

**Reply by dennisguse** [reply]
The provided accuracy is something different than HDOP. 
In Android, the GPS accuracy (i.e. location.geAccuracy()) is the radius around the measured  coordinates, so that there is a 66% chance that the real location is in this circle.
(Hope I recalled it correctly)

But we may be able to get HDOP https://stackoverflow.com/questions/6993322/how-do-i-get-the-hdop-or-vdop-values-from-the-gps-locationmanager

**Reply by dennisguse** [reply]
@brianjmurrell just checked the Google Fit link and yes: the definition for accuracy is the same as for Android.

**Reply by brianjmurrell** [reply]
So is `TrackPoint.getHorizontalAccuracy()` exactly that value that Google Fit wants then?

Even if so, I wonder what the most acceptable means of expressing that in a GPX are.  That was actually where I started and came across HDOP.

**Reply by brianjmurrell** [reply]
I suppose, lacking any existing support, by extensions or otherwise, we could add an `<opentracks:gps_accuracy>` extension.

**Reply by dennisguse** [reply]
#683

And yes that would be it.
No rocket science here.
Until now this data is not that exported at all, which is not good.

**Comment by brianjmurrell** [comment]
> But it seems that this horizontal accuracy would be the solution.

Seems not.  Even tracks that don't enable GPS have accuracy set on their points.  So a lack of accuracy cannot be used as a test for whether GPS was enabled or not.  So it seems I am back to including some kind of element or attribute (whatever the XML nomenclature is for it) to indicate whether GPS was actually enabled for the track or not.

The interesting thing is that in my GPX->Google Fit app, I made a WAG at a fixed 10m as the GPS accuracy of all of my `TrackPoint` -> `com.google.location.sample` conversions and the average actal accuracy over a GPS-enabled walk was actually 9.9m.  The WAG seems pretty good.  :-)

**Reply by dennisguse** [reply]
That one is easy: GPX does only support GPX-based TrackPoints. A TrackPoint has  mandatory latitude and longitude.
So, what OpenTracks does is: it aggregates all non GPS TrackPoint data into the subsequent GPS-based TrackPoint.
So, you will always get accuracy in GPX as long as your GPS provided it.

KML allows TrackPoints without latitude and longitude.

https://www.topografix.com/GPX/1/1/#type_wptType

**Reply by dennisguse** [reply]
@brianjmurrell sorry, I forgot to mention that beforehand...

**Reply by dennisguse** [reply]
So, the current solution would be to use the KML export.
And in practice, you can just copy KMLTrackImporter from OpenTracks (and the necessary data structures)...
